python/lsst/analysis/ap/apdb.py

```diff
@@ -314,17 +319,21 @@ def __init__(self, namespace, url="rubin@usdf-prompt-processing-dev.slac.stanfor
         super().__init__(butler=butler, instrument=instrument, **kwargs)
         self._connection_string = f"postgresql://{url}"
         self._namespace = namespace
+        self._engine = sqlalchemy.create_engine(self._connection_string)
+
+        metadata = sqlalchemy.MetaData(schema=namespace)
+        metadata.reflect()
+
+        # ensure tables don't have schema prepended
+        self._tables = {}
+        for table in metadata.tables.values():
+            self._tables[table.name] = table
 
     @property
     @contextlib.contextmanager
     def connection(self):
-        with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", message="pandas only supports SQLAlchemy connectable")
-            _connection = psycopg.connect(self._connection_string)
-            cursor = _connection.cursor()
-            cursor.execute(psycopg.sql.SQL("SET search_path TO {}").format(
-                psycopg.sql.Identifier(self._namespace)))
-            try:
-                yield _connection
-            finally:
-                _connection.close()
+        try:
+            _connection = self._engine.connect()
+            yield _connection
+        finally:
+            _connection.close()
```

Comment:
I sqlalchemy this does not guarantee that connection is actually closed. Sqlalchemy uses connection pooling, connections that are "closed" are just returned to a pool to be reused later. If you want to ensure that connection gets actually closed you need to control pooling policy by selecting NullPool pool class: https://docs.sqlalchemy.org/en/14/core/pooling.html#switching-pool-implementations

Reply:
Thanks, I've set this to use NullPool. Given that one of the uses of this will be in notebooks that might be very long lived, I do want the connection to be truly closed for postgres.